### PR TITLE
Mobile Automation Testing with Appium: An Introduction

### LINKED ISSUE #1341: Mobile Automation Testing with Appium: An Introduction
### Brief Summary: 
Appium is a popular mobile automation testing tool used for testing mobile browsers and mobile applications. This article introduces readers to this important tool that tests numerous mobile phone apps. 

### Key Takeaways: 
1.	Introduction to Appium automation testing
2.	Why Appium is widely used
3.	Features of Appium
4.	Appium architecture
5.	How Appium works
6.	Limitations of Appium
7.	Conclusion

### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)

### PR DESCRIPTION
**Pull request by mbaabuones**
### Article Checklist
**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process, quality, and style-once published, you can then submit any number of articles.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #1341 .

### CODE REVIEW

**Comment by katungi** [comment]
Hey @mbaabuones Did you delete the read.me on purpose??

**Comment by mbaabuones** [comment]
> 
> 
> Hey @mbaabuones Did you delete the read.me on purpose??

@katungi I deleted it because it's not part of the PR

**Comment by srishilesh** [comment]
@hectorkambow I would like to review this article.

File: articles/mobile-automation-testing-with-appium/index.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+#### End device
+This refers to an emulator or a mobile device that is connected to the Appium server. The test cases are executed on this device.
+
+### How Appium works?
+Appium works using the three main components of its architecture (client, server, and end device). When this tool is installed on your computer, it sets up an HTTP server that generates a REST API.
+
+This server collects command requests from the Appium client in JSON format. JSON Wire Protocol plays the role of communicating these requests. These command requests are then executed on the end devices (either Android, Windows, or iOS).
+
+The Appium server uses a test automation framework to execute requests on the user interface of end devices. For example, it uses *XCUI test* to execute commands on iOS. It uses *UI Automator* to execute requests on Android platforms (for newer API versions). After the requests have been executed, the Appium server provides HTTP responses to the client.
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:57]: A few suggestions:
- Could you brief about `XCUI test` and `UI Automator`?
- Please describe on how the test automation framework works.
- `to execute requests` Please be more specific on what requests are being executed.

File: articles/mobile-automation-testing-with-appium/index.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+#### How Appium works on Android?
+![Appium on Android](/engineering-education/articles/mobile-automation-testing-with-appium/appium-on-android.png)
+[Image Source: Edureka](https://www.edureka.co/blog/content/ver.1556540029/uploads/2019/05/Appium-on-Android-Appium-Architecture-edureka-4.png)
+
+The following is a brief sequence of how Appium works on the Android platform:
+1.The Appium client connects with the HTTP server and uses the JSON Wire Protocol for communication.
+2.The Appium server receives requests from the client component.
+3.The server connects with the mobile automation framework for Android (UI Automator or Selendroid).
+4.Android devices consist of **bootstrap.jar** that plays the role of a [TCP server](https://en.wikipedia.org/wiki/Transmission_Control_Protocol). The **bootstrap.jar** receives the test commands from the Appium server.  
+5.**Bootstrap.jar** uses the UI Automator or Selendroid to execute the action on the Android device.
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:70]: Execute what action?

File: articles/mobile-automation-testing-with-appium/index.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+6.The test results are then sent to the server, which eventually sends HTTP responses to the Appium client.
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:71]: It would be great, if you could add more info on what type of information does the HTTP request and response contain.

File: articles/mobile-automation-testing-with-appium/index.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+#### How Appium works on iOS?
+![Appium on iOS](/engineering-education/articles/mobile-automation-testing-with-appium/appium-on-ios.png)
+[Image Source: Edureka](https://www.edureka.co/blog/content/ver.1556540029/uploads/2019/05/Appium-on-iOS-Appium-Architecture-edureka-4.png)
+
+The following steps show how Appium works on the iOS platform:
+1.The Appium client connects with the HTTP server and uses the JSON Wire Protocol for communication.
+2.The Appium server receives requests from the client component.
+3.The server connects with the mobile automation framework for iOS devices (XCUI test).
+4.iOS devices consist of **bootstrap.js** that plays the role of a TCP server. **Bootstrap.js** receives the test commands from the Appium server.  
+5.**Bootstrap.js** uses the XCUI test to execute the action on the iOS device.
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:82]: Specify what action is being executed

Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:82]: `bootstrap.js`/`bootstrap.jar` and `Bootstrap.js`/`Bootstrap.jar`.
Maintain consistency throughout the article. Please make these changes for the previous topic too.

File: articles/mobile-automation-testing-with-appium/index.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+#### How Appium works on Android?
+![Appium on Android](/engineering-education/articles/mobile-automation-testing-with-appium/appium-on-android.png)
+[Image Source: Edureka](https://www.edureka.co/blog/content/ver.1556540029/uploads/2019/05/Appium-on-Android-Appium-Architecture-edureka-4.png)
+
+The following is a brief sequence of how Appium works on the Android platform:
+1.The Appium client connects with the HTTP server and uses the JSON Wire Protocol for communication.
+2.The Appium server receives requests from the client component.
+3.The server connects with the mobile automation framework for Android (UI Automator or Selendroid).
+4.Android devices consist of **bootstrap.jar** that plays the role of a [TCP server](https://en.wikipedia.org/wiki/Transmission_Control_Protocol). The **bootstrap.jar** receives the test commands from the Appium server.
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:69]: A few suggestions:
- Explain what is the role of TCP server here.
- Also, explain what does `bootstrap.jar` contain.

File: articles/mobile-automation-testing-with-appium/index.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+[Image Source: Edureka](https://www.edureka.co/blog/content/ver.1556540029/uploads/2019/05/Appium-on-iOS-Appium-Architecture-edureka-4.png)
+
+The following steps show how Appium works on the iOS platform:
+1.The Appium client connects with the HTTP server and uses the JSON Wire Protocol for communication.
+2.The Appium server receives requests from the client component.
+3.The server connects with the mobile automation framework for iOS devices (XCUI test).
+4.iOS devices consist of **bootstrap.js** that plays the role of a TCP server. **Bootstrap.js** receives the test commands from the Appium server.  
+5.**Bootstrap.js** uses the XCUI test to execute the action on the iOS device.
+6.The test results are then sent to the server, which eventually sends HTTP responses to the Appium client.
+
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:84]: Also, the working of Appium in Android/iOS could be more elaborative. The whole topic on the working of Appium is not very beginner friendly. If this article is not intended for beginners, specifying the reader level of understanding, at the beginning of the article would be important.

File: articles/mobile-automation-testing-with-appium/index.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+### Limitations of Appium
+- There are no detailed reports that provide adequate tutorial guides.
+- The tests rely on a web driver (remote), which may make them slow.
+- The support for older Android APIs is done indirectly through another library (Selendroid).
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:88]: This sentence could be rephrased like `Appium does not support older Android APIs. Alternatively, Selendroid could be used.`

File: articles/mobile-automation-testing-with-appium/index.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+1.The Appium client connects with the HTTP server and uses the JSON Wire Protocol for communication.
+2.The Appium server receives requests from the client component.
+3.The server connects with the mobile automation framework for iOS devices (XCUI test).
+4.iOS devices consist of **bootstrap.js** that plays the role of a TCP server. **Bootstrap.js** receives the test commands from the Appium server.  
+5.**Bootstrap.js** uses the XCUI test to execute the action on the iOS device.
+6.The test results are then sent to the server, which eventually sends HTTP responses to the Appium client.
+
+### Limitations of Appium
+- There are no detailed reports that provide adequate tutorial guides.
+- The tests rely on a web driver (remote), which may make them slow.
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:87]: Explain what a web driver is, and how it makes them slow.

File: articles/mobile-automation-testing-with-appium/index.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+The following steps show how Appium works on the iOS platform:
+1.The Appium client connects with the HTTP server and uses the JSON Wire Protocol for communication.
+2.The Appium server receives requests from the client component.
+3.The server connects with the mobile automation framework for iOS devices (XCUI test).
+4.iOS devices consist of **bootstrap.js** that plays the role of a TCP server. **Bootstrap.js** receives the test commands from the Appium server.  
+5.**Bootstrap.js** uses the XCUI test to execute the action on the iOS device.
+6.The test results are then sent to the server, which eventually sends HTTP responses to the Appium client.
+
+### Limitations of Appium
+- There are no detailed reports that provide adequate tutorial guides.
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:86]: Did you mean there is no/poor developer documentation? This sentence is unclear.

**Review by srishilesh — CHANGES_REQUESTED** [review]
Hi @mbaabuones ,

This is a good article. Though, I'm a beginner in mobile app development, I learnt a lot about mobile automation testing. I have requested few changes, please let me know once you are done, we can move on to the final review.

A few more suggestions:
- Add hero image
- Specify the image path correctly (You have missed out `/articles`)
- For bullet points, always use `-`, not `*`
- Remove unnecessary spaces at the end of lines.
- In conclusion, you could mention a summary of what the reader has learnt from your article. Something like [this](https://www.section.io/engineering-education/introduction-to-built-in-data-structures-in-java/#conclusion).
- Also, avoid repeating points.

File: articles/mobile-automation-testing-with-appium/index.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+- It allows the test scripts to be executed in a parallel way (simultaneously).
+- It consists of a recording tool for recording test cases and a playback tool for playing such records.
+- Various elements can be tested simultaneously (ie. emulators, real devices, and simulators).
+- It supports cloud-based testing using [testdriod](https://en.wikipedia.org/wiki/Testdroid).
+- It consists of a lively community (google group) that improves the ease and speed of troubleshooting.
+  
+### Appium architecture
+The following diagram shows the architecture of Appium:
+![Appium Architecture](/engineering-education/articles/mobile-automation-testing-with-appium/appium-architecture.jpg)
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:38]:
```suggestion
![Appium Architecture](/engineering-education/mobile-automation-testing-with-appium/appium-architecture.jpg)
```

File: articles/mobile-automation-testing-with-appium/index.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+Appium works using the three main components of its architecture (client, server, and end device). When this tool is installed on your computer, it sets up an HTTP server that generates a REST API.
+
+This server collects command requests from the Appium client in JSON format. JSON Wire Protocol plays the role of communicating these requests. These command requests are then executed on the end devices (either Android, Windows, or iOS).
+
+The Appium server uses a test automation framework to execute requests on the user interface of end devices. For example, it uses *XCUI test* to execute commands on iOS. It uses *UI Automator* to execute requests on Android platforms (for newer API versions). After the requests have been executed, the Appium server provides HTTP responses to the client.
+
+Appium works differently on iOS and Android. Let's look at how it works on these two platforms.
+
+#### How Appium works on Android?
+![Appium on Android](/engineering-education/articles/mobile-automation-testing-with-appium/appium-on-android.png)
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:62]:
```suggestion
![Appium on Android](/engineering-education/mobile-automation-testing-with-appium/appium-on-android.png)
```

File: articles/mobile-automation-testing-with-appium/index.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+The following is a brief sequence of how Appium works on the Android platform:
+1.The Appium client connects with the HTTP server and uses the JSON Wire Protocol for communication.
+2.The Appium server receives requests from the client component.
+3.The server connects with the mobile automation framework for Android (UI Automator or Selendroid).
+4.Android devices consist of **bootstrap.jar** that plays the role of a [TCP server](https://en.wikipedia.org/wiki/Transmission_Control_Protocol). The **bootstrap.jar** receives the test commands from the Appium server.  
+5.**Bootstrap.jar** uses the UI Automator or Selendroid to execute the action on the Android device.
+6.The test results are then sent to the server, which eventually sends HTTP responses to the Appium client.
+
+#### How Appium works on iOS?
+![Appium on iOS](/engineering-education/articles/mobile-automation-testing-with-appium/appium-on-ios.png)
```
Comment (srishilesh) [inline articles/mobile-automation-testing-with-appium/index.md:74]:
```suggestion
![Appium on iOS](/engineering-education/mobile-automation-testing-with-appium/appium-on-ios.png)
```

**Review by srishilesh — CHANGES_REQUESTED** [review]
Hi @mbaabuones,
Please revert the following changes. I made a mistake of adding `/articles` to the image paths (which is incorrect).

Thank you.

File: articles/mobile-automation-testing-with-appium/index.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+Appium provides tools for the automation testing of real physical objects, emulators, and simulators. Automation testing in Appium does not depend on the operating system of a mobile device. This is because its framework can translate the driver commands into Android or iOS commands that do not depend on the OS type. Instead, these commands depend on the type of mobile device.
+
+The following are some of the main types of mobile apps that can be tested using Appium automation testing:
+
+- **Native apps:** These are platform-specific apps that are developed using Android, iOS, or Windows SDKs. They operate on specific platforms. For example, either iOS or Android. These apps do not have features for web browser navigation.
+- **Hybrid apps:** These are apps that can work well on a certain platform and still have features for web browser navigation. They are a combination of mobile applications and features for web browser navigation.
+- **Web apps:** These are mobile apps that can be accessed using browsers of mobile devices (built-in).
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/mobile-automation-testing-with-appium/index.md]: I have used web browser navigation to provide a comparison of the three types of apps. For example, native apps do not provide web browser navigation while web apps allow it. Hybrid apps are simply a combination of the attributes of native apps and web apps. In the second comment, I have deleted the word 'mobile' to avoid confusion.

File: articles/mobile-automation-testing-with-appium/index.md (hunk 8 above)
Comment (mbaabuones) [reply articles/mobile-automation-testing-with-appium/index.md]: I have revised this section.

File: articles/mobile-automation-testing-with-appium/index.md (hunk 9 above)
Comment (mbaabuones) [reply articles/mobile-automation-testing-with-appium/index.md]: I have revised this section.

**Comment by mbaabuones** [comment]
> 
> 
> Hi @mbaabuones,
> Please revert the following changes. I made a mistake of adding `/articles` to the image paths (which is incorrect).
> 
> Thank you.

Alright. No problem.

**Comment by mbaabuones** [comment]
@srishilesh Thank you for your valuable suggestions. I have finished working on the changes.

**Comment by hectorkambow** [comment]
plagiarism check done

**Review by srishilesh — APPROVED** [review]
Hey @mbaabuones ,

Thanks for making the requested changes. This article looks really polished now. We can move on to the final review.
